File: Assets/Scripts/Lua/ChatSystem/ChatMsgView.lua
local GS = GS
local GF = GF
local _G, g_Env, print, os, math
    = _G, g_Env, print, os, math
local class, typeof, type, string, utf8
    = class, typeof, type, string, utf8

local _STR_ = _STR_
local SEnv = SEnv

_ENV = moduledef {}

local musicMute = false
local audioMute = false

local Class = class()

function Create(...)
    return Class(...)
end

--root:RectTransform类型
function Class:__init(view, chatPanel)
    self.view = view
    self.chatPanel = chatPanel -- 保存ChatPanel 用来统一播放音频，不自己单独播放了
    view:GetComponent(typeof(GS.LuaInitHelper)):Init(self)
    self.eventListener:Init(self)

    self.paddingAtIconSide = self.rootLayoutGroup.padding.right
    self.paddingAtOtherSide = self.rootLayoutGroup.padding.left
    self.colorAtInit = self.contentBackImage.color

    self.isPlaying = false
    -- data 信息
    self.msgData = nil

    --
    self.progressSliderRoot:SetActive(false)
    self.btn_ReSend.gameObject:SetActive(false)
end

-- data : ChatMsgData.lua类型
function Class:UpdateFromData(data)
    self.eventListener:Init(self)
    if data == nil then
        GF.logError("UpdateFromData data is nil")
        return
    end
    self.msgData = data
    -- 时间戳
    self.timeText.text = data:GetTimeStamp()
    -- 头像和布局
    self.userID = data.userID
    if data.isMine then
        self.rightHeadBG:SetActive(true)
        SEnv.AutoUpdateHeadImage(self.rightHeadImage, data.headID, data.userID)
        self.leftIHeadBG:SetActive(false)
        -- 自己发言不用显示名字
        --self.nameText.gameObject:GetComponent("RectTransform").sizeDelta = GS.Vector2(1, 30)
        self.nameText.gameObject:SetActive(false)
        --self.InfoRootLayout.childAlignment = GS.TextAnchor.UpperRight
        --self.contentBackImage.color = GS.Color(0.75, 1, 1, self.colorAtInit.a)
        self.rootLayoutGroup.childAlignment = GS.TextAnchor.UpperRight
        self.msgContentLayoutGroup.childAlignment = GS.TextAnchor.MiddleRight
        self.rootLayoutGroup.padding.right = self.paddingAtIconSide
        self.rootLayoutGroup.padding.left = self.paddingAtOtherSide
    else
        self.leftIHeadBG:SetActive(true)
        SEnv.AutoUpdateHeadImage(self.leftHeadImage, data.headID, data.userID)
        self.rightHeadBG:SetActive(false)
        --
        --self.nameText.gameObject:GetComponent("RectTransform").sizeDelta = GS.Vector2(210, 30)
        self.nameText.gameObject:SetActive(true)
        self.nameText.text = data.nickName
        --self.InfoRootLayout.childAlignment = GS.TextAnchor.UpperLeft
        --self.contentBackImage.color = self.colorAtInit
        self.rootLayoutGroup.childAlignment = GS.TextAnchor.UpperLeft
        self.msgContentLayoutGroup.childAlignment = GS.TextAnchor.MiddleLeft
        self.rootLayoutGroup.padding.right = self.paddingAtOtherSide
        self.rootLayoutGroup.padding.left = self.paddingAtIconSide
    end
    -- 设置消息条目背景
    self.contentBackImage.sprite = data.msgItemBgSpr
    -- 消息内容
    if data.audioClip ~= nil then
        self.wfDraw.gameObject:SetActive(true)
        self.text.gameObject:SetActive(false)

        self.audioSource.clip = data.audioClip
        self.wfDraw:StartWaveFormGeneration(data.audioClip)
        -- self:StartPlayback() -- 不自动播放
        self.text.text = ""
        -- 消息状态
        if data.isMine then
            self.progressSliderRoot:SetActive(not self.msgData.IsSendSusseed)
        else
            self.progressSliderRoot:SetActive(false)
        end
        self.btn_ReSend.gameObject:SetActive(false)
    else
        self.wfDraw.gameObject:SetActive(false);
        self.text.gameObject:SetActive(true);

        self.text.text = self:FixEmojiSize(data.text)
        self.audioSource.clip = nil;
    end
end

function Class:FixEmojiSize(text, fixSize)
    fixSize = fixSize or 45
    local tempText = text
    tempText = string.gsub(text, "<sprite=%d+>", function (s)
        -- print(s)
        return "<size="..fixSize..">"..s.."</size>"
    end)
    return tempText
end

function Class:StartPlayback()
    self.chatPanel:OnPlayAudioClip(self.msgData.audioClip, self.wfDraw)
end

function Class:StopPlayback()
    self.chatPanel:OnStopAudioClip(self.wfDraw)
end

function Class:OnSendSuccess()
    self.progressSliderRoot:SetActive(false)
    self.btn_ReSend.gameObject:SetActive(false)
    self.msgData.IsSendSusseed = true
    print("发送成功...")
end

function Class:OnSendFailed(err)
    SEnv.ShowHintMessage(_STR_'发送失败:'..err)
    self.progressSliderRoot:SetActive(false)
    self.btn_ReSend.gameObject:SetActive(true)
end

function Class:OnUpdateProgress(progress)
    self.progressSlider.value = progress
    -- print("发送进度:", progress*100)
end

function Class:OnDestroy()
    -- print("ChatMsgViewItem OnDestroy.............")
end

-- function Class:OnEnable()
--     print("ChatMsgView OnEnable.............")
-- end

function Class:On_btn_ReSend_Event(btn_ReSend)
    if self.OnReSend then
        self.OnReSend(self.timestampSec)
    end
end

function Class:On_btn_PlaySound_Event(btn_PlaySound)
    self:StartPlayback()
end


return _ENV

File: Assets/ForReBuild/Scripts/ChatSystem/ChatMsgView.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using TMPro;
using UnityEngine.UI;

namespace SP
{

    public class ChatMsgData
    {
        public int timestampSec;    // 时间戳（秒数）
        public int userID;          // 玩家ID(用来获取发送消息的用户信息)
        public string text;         // 消息文本
        public bool isMine;
        public AudioClip clip;      // 消息音频
        public Sprite iconSpr;         // 头像
    }

    public class ChatMsgView : MonoBehaviour
    {
        // 计时原点
        public readonly DateTime EPOCH_START_TIME = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
        //
        public Image leftIcon, rightIcon;   // 左右两边头像（可替换成按钮打开用户信息页）
        public TMP_Text timeText;       // 时间戳
        public TMP_Text text;           // 消息文本
        public AudioSource audioSource; // 消息音频
        public WaveFormDraw wfDraw;     // 音频波形绘制
        public Action onClick;          // 点击事件

        public ContentSizeFitter sizeFitter;
        public VerticalLayoutGroup rootLayoutGroup;
        public VerticalLayoutGroup msgContentLayoutGroup;
        public Image contentBackImage;

        int paddingAtIconSide, paddingAtOtherSide;
        Color colorAtInit;
        int userID;
        bool isMine;  // 是否为自己发的消息
        // 音频消息是否在播放
        bool isPlaying = false;

        private void Awake()
        {
            paddingAtIconSide = rootLayoutGroup.padding.right;
            paddingAtOtherSide = rootLayoutGroup.padding.left;
            //sizeFitter.enabled = false;
            colorAtInit = contentBackImage.color;
        }

        private void OnDestroy() {
            onClick = null;
        }

        public void UpdateFromData(ChatMsgData data)
        {
            if (data == null)
            {
                Debug.LogError("data is null ");
                return;
            }
            // 时间戳
            DateTime dtDataTime = EPOCH_START_TIME.AddSeconds(data.timestampSec).ToLocalTime();
            timeText.text = dtDataTime.ToString("HH:mm:ss");
            // 头像和布局 
            userID = data.userID;
            isMine = data.isMine;
            if(isMine)
            {
                rightIcon.gameObject.SetActive(true);
                rightIcon.sprite = data.iconSpr;
                leftIcon.gameObject.SetActive(false);
                //
                contentBackImage.color = new Color(0.75f, 1f, 1f, colorAtInit.a);
                rootLayoutGroup.childAlignment = msgContentLayoutGroup.childAlignment = TextAnchor.MiddleRight;
                rootLayoutGroup.padding.right = paddingAtIconSide;
                rootLayoutGroup.padding.left = paddingAtOtherSide;
            }
            else
            {
                leftIcon.gameObject.SetActive(true);
                leftIcon.sprite = data.iconSpr;
                rightIcon.gameObject.SetActive(false);
                //
                contentBackImage.color = colorAtInit;
                rootLayoutGroup.childAlignment = msgContentLayoutGroup.childAlignment = TextAnchor.MiddleLeft;
                rootLayoutGroup.padding.right = paddingAtOtherSide;
                rootLayoutGroup.padding.left = paddingAtIconSide;
            }

            // 消息内容
            if(data.clip != null)   // 音频消息
            {
                wfDraw.gameObject.SetActive(true);
                text.gameObject.SetActive(false);

                audioSource.clip = data.clip;
                wfDraw.StartWaveFormGeneration(data.clip);
                onClick = () => {
                    // play audioclip
                    if (isPlaying)
                        StopPlayback();
                    else
                        StartPlayback();

                };
                text.text = "";
            }
            else
            {   
                wfDraw.gameObject.SetActive(false);
                text.gameObject.SetActive(true);

                text.text = data.text;
                onClick = null;
                audioSource = null;
            }
        }


        private void Update()
        {
            if (isPlaying) // 回放
            {
                if (audioSource == null || audioSource.clip == null)
                    return;
                wfDraw.playbackSli.value = audioSource.timeSamples * audioSource.clip.channels;
                if (audioSource.isPlaying == false)
                {
                    StopPlayback();
                }
            }
        }

        public void StartPlayback()
        {
            if (isPlaying)
                return;

            isPlaying = true;
            audioSource.Play();
            print("回放开始....");
        }

        public void StopPlayback()
        {
            if (isPlaying == false)
                return;
            isPlaying = false;
            audioSource.Stop();
            wfDraw.playbackSli.value = 0;
            print("回放结束....");
        }

    }
}

